### PR TITLE
feat(python): extend `filter` capabilities with new support for `*args` predicates, `**kwargs` constraints, and chained boolean masks

### LINKED ISSUE #11642: Add .locate method as an affordance for filtering dataframes
### Description

I found the following method extremally convenient in my code:
```python
def _locate(df:pl.DataFrame, **kwargs)->pl.DataFrame:
    '''Used to quickly find rows via equality checks.
    
    Example:
    # This
    df.locate(year=2020, quarter=2, metric='Sales')
    
    # is equivalent to 
    df.filter(pl.col('year')==2020).filter(pl.col('quarter')==2).filter(pl.col('metric')=='Sales')
    '''
            
    df_ans = df.filter(pl.all_horizontal([pl.col(k)==v for k,v in kwargs.items()]))
    
    return df_ans

pl.DataFrame.locate = _locate
```

I usualy would not recommend these methods be added to polars, but I was showing a friend the library on a clean installation and realized how much I would miss it if it were gone.

It allows you to write code like this:

```python
# Get the SignOn events for user #9
df_selected = df.locate(userId=9, eventType='SignOn')

# Equivalent to 
df_selected = df.filter(pl.col('userId')==9 & pl.col('eventType')=='SignOn')

```
Incidentally it also supports expressions:

```python
# Get the most recent SignOn date for user #9
df_selected = df.locate(userId=9, eventType='SignOn',eventDate=pl.col('eventDate').max().over('userId','eventType'))   
```

Thoughts? Is the task of retrieving a subset of rows via a few columns common enough to merit a shorthand?

### PR DESCRIPTION
**Pull request by alexander-beedie**
Closes #11642, supersedes #11674.

Ended up going a little deeper down the rabbit hole so moved it into my own branch to finish off as I was having issues rebasing / force-pushing to a branch I don't own (some sort of credentials/token issue? 🤷); kudos to @mcrumiller for getting things started with the initial implementation!

## Miscellaneous

* Turns out we can cleanly disambiguate new/old usage of the `predicate` keyword, so only warns on old-style usage.
* Noticed a fast-path for recognising numpy-based boolean masks (and int sequences), so integrated this too.
* The `parse_as_expression` docstring claimed it had a "wrap" param but it didn't - so I added it (and used it) ;)

## New Features

Adds support for:
* `*predicates` (one _or more_ predicates as positional arguments).
* `**constraints` (`col == value` matches as keyword arguments).
* resolves _chains_ of boolean masks (previously only allowed one).

## Examples
More than one predicate, now we have a positional args option.
Implicitly combines predicates with `all_horizontal`:
```python
df.filter( 
    pl.col("a") >= 0, 
    pl.col("b").is_not_null(),
    pl.col("").str.starts_with("!"),
 )
```
Splat predicates assembled elsewhere straight into filter without having to combine them yourself:
```python
df.filter( *conditions )
```
Constrain rows using `**kwargs` format (essentially the same API that we offer in `with_columns`):
```python
df.filter( 
    x = 0,
    y = "xxx",
    z = date.today(),
)
```
Mix and match the two formats freely:
```python
df.filter( *conditions, value=0, start_date=date.today() )
```
Easily filter against a Pydantic/FastAPI `Model` object:
```python
from pydantic import BaseModel
from datetime import date
import polars as pl

class DataModel(BaseModel):
    x: date
    y: str
    z: int
    
model_data = [
    DataModel( x=date.today(), y="aa", z=123 ),
    DataModel( x=date.today(), y="bb", z=456 ),
    DataModel( x=date.today(), y="cc", z=789 ),
]

df = pl.DataFrame( data=model_data )
# shape: (3, 3)
# ┌────────────┬─────┬─────┐
# │ x          ┆ y   ┆ z   │
# │ ---        ┆ --- ┆ --- │
# │ date       ┆ str ┆ i64 │
# ╞════════════╪═════╪═════╡
# │ 2023-10-14 ┆ aa  ┆ 123 │
# │ 2023-10-14 ┆ bb  ┆ 456 │
# │ 2023-10-14 ┆ cc  ┆ 789 │
# └────────────┴─────┴─────┘

df.filter( **model_data[1].model_dump() )
# shape: (1, 3)
# ┌────────────┬─────┬─────┐
# │ x          ┆ y   ┆ z   │
# │ ---        ┆ --- ┆ --- │
# │ date       ┆ str ┆ i64 │
# ╞════════════╪═════╪═════╡
# │ 2023-10-14 ┆ bb  ┆ 456 │
# └────────────┴─────┴─────┘
```
Chain boolean masks (previously only allowed one per `filter` invocation):
```python
mask1 = [True, False, True]
mask2 = [False, True, True]

df.filter( mask1, mask2 )
# shape: (1, 3)
# ┌────────────┬─────┬─────┐
# │ x          ┆ y   ┆ z   │
# │ ---        ┆ --- ┆ --- │
# │ date       ┆ str ┆ i64 │
# ╞════════════╪═════╪═════╡
# │ 2023-10-14 ┆ cc  ┆ 789 │
# └────────────┴─────┴─────┘
```

### CODE REVIEW

File: py-polars/polars/dataframe/frame.py [hunk 1]
```diff
@@ -3899,7 +3899,8 @@ def insert_at_idx(self, index: int, series: Series) -> Self:
 
     def filter(
         self,
-        predicate: (Expr | str | Series | list[bool] | np.ndarray[Any, Any] | bool),
+        *predicates: IntoExpr | list[bool] | np.ndarray[Any, Any],
```
Comment (stinodego) [inline py-polars/polars/dataframe/frame.py:3902]: We now have `IntoExprColumn` for cases like this - this way, we don't allow things like datetimes:

```suggestion
        *predicates: IntoExprColumn | bool | list[bool] | np.ndarray[Any, Any],
```

File: py-polars/polars/dataframe/frame.py [hunk 2]
```diff
@@ -3959,12 +3960,38 @@ def filter(
...
+        │ i64 ┆ i64 ┆ str │
+        ╞═════╪═════╪═════╡
+        │ 1   ┆ 6   ┆ a   │
+        └─────┴─────┴─────┘
+
+        """
         return (
-            self.lazy().filter(predicate).collect(_eager=True)  # type: ignore[arg-type]
+            self.lazy()
+            .filter(*predicates, **constraints)  # type: ignore[arg-type]
```
Comment (stinodego) [inline py-polars/polars/dataframe/frame.py:3993]: Looks like the types of `LazyFrame.filter` need to be updated to match `DataFrame.filter`

File: py-polars/polars/lazyframe/frame.py [hunk 3]
```diff
@@ -2602,11 +2634,60 @@ def filter(self, predicate: IntoExpr) -> Self:
...
+                is_mask := is_bool_sequence(p)
+            ):
+                p = constraints.pop("predicate")
+                warnings.warn(
+                    "`filter` no longer takes a 'predicate' parameter.\n"
+                    "To silence this warning you should omit the keyword and pass "
+                    "as a positional argument instead.",
+                    DeprecationWarning,
+                    stacklevel=find_stacklevel(),
+                )
```
Comment (stinodego) [inline py-polars/polars/lazyframe/frame.py:2658]: We have the `issue_deprecation_warning` util for this now!

File: py-polars/polars/lazyframe/frame.py [hunk 4]
```diff
@@ -2602,11 +2634,60 @@ def filter(self, predicate: IntoExpr) -> Self:
...
+                if is_mask:
+                    boolean_masks.append(pl.Series(p, dtype=Boolean))
+                else:
+                    all_predicates += (p,)
```
Comment (stinodego) [inline py-polars/polars/lazyframe/frame.py:2662]: Just use `append` here, like on line 2643?
```suggestion
                    all_predicates.append(p)
```

File: py-polars/polars/lazyframe/frame.py [hunk 5]
```diff
@@ -2602,11 +2634,60 @@ def filter(self, predicate: IntoExpr) -> Self:
         └─────┴─────┴─────┘
 
         """
-        if isinstance(predicate, list):
-            predicate = pl.Series(predicate)
+        # note: identify masks separately from predicates
+        all_predicates, boolean_masks = [], []
+        for p in predicates:
```
Comment (stinodego) [inline py-polars/polars/lazyframe/frame.py:2639]: All our `*args` functions support passing a single list as well, like `df.filter([pl.col('a') == 2, pl.col('b') == 3])`. I don't believe the current parsing logic supports this - we should add that!

File: py-polars/polars/utils/_parse_expr_input.py [hunk 6]
```diff
@@ -77,6 +77,7 @@ def parse_as_expression(
     *,
     str_as_lit: bool = False,
     structify: bool = False,
+    wrap: bool = False,
```
Comment (stinodego) [inline py-polars/polars/utils/_parse_expr_input.py:80]: I specifically chose not to add this - just call `wrap_expr` on the result.

And when you find you need this option, it means you're doing things in Python that should be done in Rust.

I'll leave it alone for now, but really all the logic you implemented here should probably be done in Rust.

**Review by stinodego — CHANGES_REQUESTED** [review]
I wonder if we can't just use `parse_as_list_of_expressions` here? There seems to be a lot of duplicated parsing logic here.

File: py-polars/polars/lazyframe/frame.py (hunk 3 above)
Comment (alexander-beedie) [reply py-polars/polars/lazyframe/frame.py:2658]: Ah, nice - I must have missed that being introduced  ;)

File: py-polars/polars/lazyframe/frame.py (hunk 4 above)
Comment (alexander-beedie) [reply py-polars/polars/lazyframe/frame.py:2662]: Ahh! Good spot - in an earlier draft I was appending onto the 'predicates' tuple where yo can't append 😅

File: py-polars/polars/utils/_parse_expr_input.py (hunk 6 above)
Comment (alexander-beedie) [reply py-polars/polars/utils/_parse_expr_input.py:80]: In that case just need to remove the extraneous docstring entry (odd that one of our many lints didn't catch that)

**Comment by alexander-beedie** [comment]
> I wonder if we can't just use `parse_as_list_of_expressions` here?

We can't, as the boolean masks need to be identified separately; putting that logic into `parse_as_list_of_expressions` for a single unusual use-case would have been a bit peculiar.

File: py-polars/polars/lazyframe/frame.py (hunk 5 above)
Comment (alexander-beedie) [reply py-polars/polars/lazyframe/frame.py:2653]: Done 👌

**Review by stinodego — APPROVED** [review]
Great stuff, Alex!

**Pull request merged**